File: src/binding/csharp/src/ZVec/ZVec.cs
// Copyright 2025-present the zvec project
using System;
using System.Runtime.InteropServices;

namespace ZVec
{
    public class ZVecException : Exception
    {
        public int Code { get; }
        public ZVecException(int code, string message) : base(message) { Code = code; }
    }

    internal static class StatusHelper
    {
        internal static void Check(IntPtr status)
        {
            if (status == IntPtr.Zero) return;
            int code = NativeMethods.zvec_status_code(status);
            string msg = Marshal.PtrToStringUTF8(NativeMethods.zvec_status_message(status)) ?? "Unknown error";
            NativeMethods.zvec_status_destroy(status);
            throw new ZVecException(code, msg);
        }
    }

    public class Schema : IDisposable
    {
        internal IntPtr Handle { get; private set; }

        public Schema(string name)
        {
            Handle = NativeMethods.zvec_schema_create(name);
        }

        public void AddField(string name, DataType type, uint dimension = 0)
        {
            StatusHelper.Check(NativeMethods.zvec_schema_add_field(Handle, name, (int)type, dimension));
        }

        public void Dispose()
        {
            if (Handle != IntPtr.Zero) { NativeMethods.zvec_schema_destroy(Handle); Handle = IntPtr.Zero; }
        }
    }

    public class Doc : IDisposable
    {
        internal IntPtr Handle { get; private set; }
        private bool _owns;

        public Doc()
        {
            Handle = NativeMethods.zvec_doc_create();
            _owns = true;
        }

        internal Doc(IntPtr handle)
        {
            Handle = handle;
            _owns = true;
        }

        public void SetPK(string pk) => NativeMethods.zvec_doc_set_pk(Handle, pk);

        public string PK()
        {
            IntPtr ptr = NativeMethods.zvec_doc_pk(Handle);
            return Marshal.PtrToStringUTF8(ptr) ?? "";
        }

        public void SetString(string field, string value) =>
            StatusHelper.Check(NativeMethods.zvec_doc_set_string(Handle, field, value));

        public void SetInt32(string field, int value) =>
            StatusHelper.Check(NativeMethods.zvec_doc_set_int32(Handle, field, value));

        public void SetVector(string field, float[] vector) =>
            StatusHelper.Check(NativeMethods.zvec_doc_set_float_vector(Handle, field, vector, (uint)vector.Length));

        public float Score() => NativeMethods.zvec_doc_score(Handle);

        public void Dispose()
        {
            if (_owns && Handle != IntPtr.Zero) { NativeMethods.zvec_doc_destroy(Handle); Handle = IntPtr.Zero; }
        }
    }

    public class Collection : IDisposable
    {
        private IntPtr _handle;

        public Collection(string path, Schema schema)
        {
            StatusHelper.Check(NativeMethods.zvec_collection_create_and_open(path, schema.Handle, out _handle));
        }

        public Collection(string path)
        {
            StatusHelper.Check(NativeMethods.zvec_collection_open(path, out _handle));
        }

        public void Upsert(Doc[] docs)
        {
            IntPtr[] ptrs = Array.ConvertAll(docs, d => d.Handle);
            StatusHelper.Check(NativeMethods.zvec_collection_upsert(_handle, ptrs, (nuint)ptrs.Length));
        }

        public Doc[] Fetch(string[] pks)
        {
            IntPtr[] pkPtrs = new IntPtr[pks.Length];
            for (int i = 0; i < pks.Length; i++)
                pkPtrs[i] = Marshal.StringToCoTaskMemUTF8(pks[i]);

            try
            {
                StatusHelper.Check(NativeMethods.zvec_collection_fetch(_handle, pkPtrs, (nuint)pks.Length, out IntPtr list));
                nuint size = NativeMethods.zvec_doc_list_size(list);
                Doc[] results = new Doc[(int)size];
                for (nuint i = 0; i < size; i++)
                    results[(int)i] = new Doc(NativeMethods.zvec_doc_list_get(list, i));
                NativeMethods.zvec_doc_list_destroy(list);
                return results;
            }
            finally
            {
                foreach (var p in pkPtrs) Marshal.FreeCoTaskMem(p);
            }
        }

        public void Flush() => StatusHelper.Check(NativeMethods.zvec_collection_flush(_handle));

        public ulong Stats()
        {
            StatusHelper.Check(NativeMethods.zvec_collection_get_stats(_handle, out IntPtr stats));
            ulong count = NativeMethods.zvec_stats_total_docs(stats);
            NativeMethods.zvec_stats_destroy(stats);
            return count;
        }

        public void DestroyPhysical() => StatusHelper.Check(NativeMethods.zvec_collection_destroy_physical(_handle));

        public void Dispose()
        {
            if (_handle != IntPtr.Zero) { NativeMethods.zvec_collection_destroy(_handle); _handle = IntPtr.Zero; }
        }
    }
}

File: src/binding/swift/Sources/ZVec/ZVec.swift
// Copyright 2025-present the zvec project
@_exported import CZVec
import Foundation

public enum ZVecError: Error {
    case notFound
    case alreadyExists
    case invalidArgument
    case permissionDenied
    case internalError
    case unknown(String)
}

func mapError(_ status: OpaquePointer?) throws {
    guard let status = status else { return }
    defer { zvec_status_destroy(status) }

    let code = zvec_status_code(status)
    let msg = String(cString: zvec_status_message(status))

    switch code {
    case ZVEC_NOT_FOUND: throw ZVecError.notFound
    case ZVEC_ALREADY_EXISTS: throw ZVecError.alreadyExists
    case ZVEC_INVALID_ARGUMENT: throw ZVecError.invalidArgument
    case ZVEC_PERMISSION_DENIED: throw ZVecError.permissionDenied
    case ZVEC_INTERNAL_ERROR: throw ZVecError.internalError
    default: throw ZVecError.unknown(msg)
    }
}

public class Schema {
    var ptr: OpaquePointer

    public init(name: String) {
        self.ptr = zvec_schema_create(name)!
    }

    deinit {
        zvec_schema_destroy(ptr)
    }

    public func addField(name: String, type: zvec_data_type_t, dimension: UInt32 = 0) throws {
        let status = zvec_schema_add_field(ptr, name, type, dimension)
        try mapError(status)
    }
}

public class Doc {
    var ptr: OpaquePointer

    public init() {
        self.ptr = zvec_doc_create()!
    }

    internal init(ptr: OpaquePointer) {
        self.ptr = ptr
    }

    deinit {
        zvec_doc_destroy(ptr)
    }

    public func setPK(_ pk: String) {
        zvec_doc_set_pk(ptr, pk)
    }

    public func pk() -> String {
        return String(cString: zvec_doc_pk(ptr)!)
    }

    public func setString(field: String, value: String) throws {
        let status = zvec_doc_set_string(ptr, field, value)
        try mapError(status)
    }

    public func setInt32(field: String, value: Int32) throws {
        let status = zvec_doc_set_int32(ptr, field, value)
        try mapError(status)
    }

    public func setVector(field: String, vector: [Float]) throws {
        let status = zvec_doc_set_float_vector(ptr, field, vector, UInt32(vector.count))
        try mapError(status)
    }
}

public class Collection {
    var ptr: OpaquePointer

    public init(path: String, schema: Schema) throws {
        var outPtr: OpaquePointer?
        let status = zvec_collection_create_and_open(path, schema.ptr, &outPtr)
        try mapError(status)
        self.ptr = outPtr!
    }

    public init(path: String) throws {
        var outPtr: OpaquePointer?
        let status = zvec_collection_open(path, &outPtr)
        try mapError(status)
        self.ptr = outPtr!
    }

    deinit {
        zvec_collection_destroy(ptr)
    }

    public func upsert(docs: [Doc]) throws {
        var ptrs = docs.map { Optional($0.ptr) }
        let status = zvec_collection_upsert(ptr, &ptrs, ptrs.count)
        try mapError(status)
    }

    public func fetch(pks: [String]) throws -> [Doc] {
        var pkPtrs = pks.map { ($0 as NSString).utf8String }
        var listPtr: OpaquePointer?
        let status = zvec_collection_fetch(ptr, &pkPtrs, pkPtrs.count, &listPtr)
        try mapError(status)

        let size = zvec_doc_list_size(listPtr)
        var results: [Doc] = []
        for i in 0..<size {
            let docPtr = zvec_doc_list_get(listPtr, i)
            results.append(Doc(ptr: docPtr!))
        }
        zvec_doc_list_destroy(listPtr)
        return results
    }

    public func flush() throws {
        try mapError(zvec_collection_flush(ptr))
    }

    public func stats() throws -> UInt64 {
        var statsPtr: OpaquePointer?
        try mapError(zvec_collection_get_stats(ptr, &statsPtr))
        let count = zvec_stats_total_docs(statsPtr)
        zvec_stats_destroy(statsPtr)
        return count
    }
}
